### ISSUE TITLE
Optional Support for Dropwizard JDBI broken 

### ISSUE THREAD

**Issue by pandaadb** [dropwizard/dropwizard#1297]
Hi,

I recently noticed there is a bug in the optional handling for JDBi and joda DateTime objects. There is a serialisation issue that prevents any Optional datetimes to be inserted into a table. 

I created following unittest to demonstrate the problem:
```java
public class TestJDBIOptional {

	protected static DBI dbi;
	protected static Environment env;

	static DataSourceFactory getDataSourceFactory() {
		DataSourceFactory dataSourceFactory = new DataSourceFactory();
		dataSourceFactory.setDriverClass("org.h2.Driver");
		dataSourceFactory.setUrl("jdbc:h2:mem:testDb");
		dataSourceFactory.setUser("sa");
		dataSourceFactory.setPassword("");
		return dataSourceFactory;
	}

	String createTable = "CREATE TABLE IF NOT EXISTS test ( id varchar(255) primary key, "
			+ " time timestamp);";

	@Before
	public void setupTests() throws IOException {
		dbi.open().createStatement(createTable).execute();
	}

	@BeforeClass
	public static void setup() {
		env = new Environment("test-env", Jackson.newObjectMapper(), null, new MetricRegistry(), null);
		dbi = new DBIFactory().build(env, getDataSourceFactory(), "test");
		dbi.registerArgumentFactory(new JodaDateTimeArgumentFactory());
		dbi.registerMapper(new JodaDateTimeMapper(Optional.absent()));
	}
	
	@Test
	public void testInsert() {
		JDBIDao onDemand = dbi.onDemand(JDBIDao.class);
		onDemand.insert(Optional.of(DateTime.now()));
	}

	@Test 
	public void testAbsent() {
		JDBIDao onDemand = dbi.onDemand(JDBIDao.class);
		onDemand.insert(Optional.absent());
	}
	
	public interface JDBIDao {
		 
		@SqlUpdate("INSERT INTO test VALUES ('test', :ts)")
		abstract int insert(@Bind("ts") Optional<DateTime> dt);
	}
}
```

**Comment by arteam** [comment]
Thanks for reporting this issue. You would need to register own argument factory to combine optional and JodaTime's `DateTime` arguments. Something similar to this:

````java
import com.google.common.base.Optional;
import org.joda.time.DateTime;
import org.skife.jdbi.v2.StatementContext;
import org.skife.jdbi.v2.tweak.Argument;
import org.skife.jdbi.v2.tweak.ArgumentFactory;

import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.sql.Types;


public class OptionalJodaTimeArgumentFactory implements ArgumentFactory<Optional<DateTime>> {

    @Override
    public boolean accepts(final Class<?> expectedType,
                           final Object value,
                           final StatementContext ctx) {
        if (value instanceof Optional) {
            Optional optional = (Optional) value;
            if (optional.isPresent()) {
                return optional.get().getClass().equals(DateTime.class);
            }
        }
        return false;
    }

    @Override
    public Argument build(final Class<?> expectedType,
                          final Optional<DateTime> value,
                          final StatementContext ctx) {
        return new Argument() {
            @Override
            public void apply(int position, PreparedStatement statement, StatementContext ctx) throws SQLException {
                if (value.isPresent()) {
                    statement.setTimestamp(position, new Timestamp(value.get().getMillis()));
                } else {
                    statement.setNull(position, Types.TIMESTAMP);
                }
            }
        };
    }
}
````

````java
dbi = new DBIFactory().build(env, getDataSourceFactory(), "test");
dbi.registerArgumentFactory(new OptionalJodaTimeArgumentFactory());
````

It would probably work for this specific case, but I am not sure for the general. Will need some digging and testing.

**Comment by joschi** [comment]
Solved in #1304.
